Add `ChannelReady` event

tnull opened the pull request:
Closes #1394. 

This adds a `ChannelReady` event that will be emitted as soon as a new
channel becomes usable, i.e., after both sides have sent `channel_ready`.

tnull commented:
Draft for now, since tests are still failing.

tnull commented on lightning/src/util/events.rs:
```diff
@@ -505,6 +505,27 @@ pub enum Event {
+	/// be used. This event is emitted either when the funding transaction has been confirmed
+	/// on-chain, or, in case of a 0conf channel, when both parties have confirmed the channel
+	/// establishment.
+	ChannelEstablished {
```
Not 100% sure about the name, since `ChannelEstablished` is somewhat confusing since the event should be emitted only once on channel establishment, not on reestablishment, etc. I thought about naming this `ChannelReady` (could also be confusing for people knowing about LN internals) or `ChannelConfirmed` (doesn't really work for 0conf channels..) before. So, open for further suggestions.

valentinewallace replied:
Maybe `ChannelUsable` is more clear? It feels like that's what users would care about

valentinewallace replied:
I think from a public API perspective it isn't used anywhere else, except in `usable_channels`, which is a nice alignment. No strong feelings here though

valentinewallace replied:
Oh, actually I see your point. Maybe it is confusing so feel free to disregard

tnull replied:
So, given that the new event should be the counterpart to `ChannelClosed`, could we just name it `ChannelOpened`? Might also be a bit in conflict with `ChannelOpenRequest`, but seems the most straightforward?

ViktorTigerstrom replied:
Perhaps `ChannelOperational` could work? We've used the word "operational" for some error messages when receiving htlc messages too early.

tnull replied:
Went with `ChannelOpened` for now. Let me know if there anyone has strong feelings against that.

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4658,6 +4658,14 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 						});
 					}
+
+					let mut pending_events = self.pending_events.lock().unwrap();
```
Also not 100% sure whether I covered all the cases where we should emit this event. Two feels like I'm missing something. Will have a further look.

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4658,6 +4658,14 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 					}
+
+					let mut pending_events = self.pending_events.lock().unwrap();
+					pending_events.push(events::Event::ChannelOpened {
```
I agree, we need to take care of retransmissions. I assume we would track that in the `Channel` itself?

Also, do I understand correctly that `ChannelState::ChannelFunded` is by now the equivalent of `ChannelState::OurChannelReady` && `ChannelState::TheirChannelReady`, i.e., also used for 0conf? Should it then possibly be renamed to `ChannelState::ChannelReady`?

tnull commented:
Rebased on main and reworked the approach: a `bool` is now bubbled up from `Channel::check_get_channel_ready` along side the `msgs::ChannelReady` and used in `ChannelManager` to decide whether to `emit_channel_ready!`.

In the second commit I included the discussed renaming of `ChannelState::ChannelFunded` to `ChannelState::ChannelReady`, and since all three align nicely then, I now reconsidered and named the event `Event::ChannelReady` after all.

Finally, the third commit includes a minor fix for unused import warnings.

Tests currently still failing, looking into it.

tnull commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4594,6 +4610,10 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 				if let Some(msg) = channel_ready {
 					send_channel_ready!(channel_state.short_to_chan_info, channel_state.pending_msg_events, chan, msg);
 				}
+				if emit_channel_ready_event {
```
Not sure I can follow entirely. IIUC, this would be an issue if a channel monitor update would be pending/in progress? But the bool would only ever be emitted when this would not be the case, as of
https://github.com/tnull/rust-lightning/blob/25269755c9a7ac82ef101f1b76731bf61bddb433/lightning/src/ln/channel.rs#L4937-L4946

But probably I'm misunderstanding something?

tnull replied:
Alright, now not bothering in funding methods as of https://github.com/lightningdevkit/rust-lightning/pull/1743/commits/08a72b8d1f1cb473dc40060da5d6108c223496e3.

tnull commented:
Fixed tests and rebased on main.

tnull commented on lightning/src/util/events.rs:
```diff
@@ -858,6 +882,15 @@ impl Writeable for Event {
+					(0, channel_id, required),
+					(1, user_channel_id, required),
+					(2, counterparty_node_id, required),
+					(3, channel_type, required),
```
Good point. Fixed.

tnull commented on lightning/src/ln/channel.rs:
```diff
@@ -4598,6 +4619,16 @@ impl<Signer: Sign> Channel<Signer> {
+	// Checks whether we previously had emitted a `ChannelReady` event and sets the value if not.
+	pub(crate) fn should_emit_channel_ready_event(&mut self) -> bool {
+		if self.is_usable() && !self.channel_ready_event_emitted {
+			self.channel_ready_event_emitted = true;
```
The intention was to ensure we'd never emit the event twice. But you're probably right that we should rather emit a duplicate event than never emitting an event if something goes wrong. I now (re)introduced a `set_channel_ready_event_emitted` method which is called in the macro.

dunxen commented on lightning-background-processor/src/lib.rs:
```diff
@@ -988,9 +988,12 @@ mod tests {
-		let event_handler = move |event: &Event| {
-			sender.send(handle_funding_generation_ready!(event, channel_value)).unwrap();
+		let event_handler = move |event: &Event| match event {
+			Event::FundingGenerationReady{ .. } => sender.send(handle_funding_generation_ready!(event, channel_value)).unwrap(),
```
nit: 
```suggestion
			Event::FundingGenerationReady { .. } => sender.send(handle_funding_generation_ready!(event, channel_value)).unwrap(),
```

tnull commented:
I now reverted the 'bubble-up' approach after all since all required checks are conducted in the `should_emit_channel_ready_event` and `send_channel_ready!`, which makes this much less intrusive. 

Squashed changes.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -736,6 +736,9 @@ pub(super) struct Channel<Signer: Sign> {
 	outbound_scid_alias: u64,
+
+	// We track whether we already emitted a `ChannelReady` event.
+	channel_ready_event_emitted: bool,
```
Is it a big deal if this is race-y? Would that ever happen?

tnull replied:
I think it would matter, as we may emit the event multiple times, which in turn could trigger some logic of the user depending on it. It shouldn't happen however, as we hold the lock on the channel for the duration of `emit_channel_ready_event!`. But, I now also moved taking and dropping the lock into the macro to ensure we really only setting it emitted after we pushed to the event queue and dropped its lock.

dunxen reviewed (commented):
Looks good. Just noticed we use "0conf" and "0-conf" in comments throughout the project. The hyphenated one looks "more correct". No biggie.

tnull commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -407,9 +407,26 @@ pub fn create_chan_between_nodes<'a, 'b, 'c, 'd>(node_a: &'a Node<'b, 'c, 'd>, n
 
+#[macro_export]
+#[cfg(any(test, feature = "_bench_unstable", feature = "_test_utils"))]
+macro_rules! expect_channel_ready_event {
```
Fixed in https://github.com/lightningdevkit/rust-lightning/pull/1743/commits/149a5749d13bae16687e98393de6f66b34e0fc16.

tnull commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -1981,7 +1981,6 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, OM: Deref, L: Deref, CM
 	// broadcast_node_announcement panics) of the maximum-length addresses would fit in a 64KB
 	// message...
 	const HALF_MESSAGE_IS_ADDRS: u32 = ::core::u16::MAX as u32 / (NetAddress::MAX_LEN as u32 + 1) / 2;
-	#[deny(const_err)]
```
No, it's has been warn-by-default for a while now, so specifying this is superfluous. Plus, it recently was even made a hard error, so this will deny-by-default going forward

tnull replied:
Well, it starts emitting annoying warnings? But alright, will drop the commit then for now. 🤷‍♂️

tnull replied:
Alright, dropped the commit for now, may reintroduce once warnings showing up in stable.

tnull commented on lightning/src/ln/payment_tests.rs:
```diff
@@ -499,7 +500,6 @@ fn do_retry_with_no_persist(confirm_before_reload: bool) {
 	let bs_htlc_claim_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().split_off(0);
 	assert_eq!(bs_htlc_claim_txn.len(), 1);
 	check_spends!(bs_htlc_claim_txn[0], as_commitment_tx);
-	expect_payment_forwarded!(nodes[1], nodes[0], nodes[2], None, false, false);
```
Ah, we want to move this because the `PaymentForwarded` event is emitted by the `handle_update_fulfill_htlc` already and then the `expect_channel_ready_event` call in `create_announced_chan_between_nodes` fails due to 2 events being queued.

tnull commented on lightning/src/util/events.rs:
```diff
@@ -1138,6 +1171,29 @@ impl MaybeReadable for Event {
+					Ok(Some(Event::ChannelReady {
+						channel_id,
+						user_channel_id,
+						counterparty_node_id: counterparty_node_id_opt.unwrap(),
```
Fixed with https://github.com/lightningdevkit/rust-lightning/pull/1743/commits/40a170eb0e93ac4fa802b9b77d5be0dd8edf266a.

tnull commented on lightning/src/ln/channel.rs:
```diff
@@ -6666,6 +6688,8 @@ impl<'a, Signer: Sign, K: Deref> ReadableArgs<(&'a K, u32)> for Channel<Signer>
 			// Later in the ChannelManager deserialization phase we scan for channels and assign scid aliases if its missing
 			outbound_scid_alias: outbound_scid_alias.unwrap_or(0),
 
+			channel_ready_event_emitted: channel_ready_event_emitted.unwrap_or(false),
```
Fixed in https://github.com/lightningdevkit/rust-lightning/pull/1743/commits/2ce2b155c8c99f529362ffa2802bb242a8f22519.

tnull commented on lightning/src/util/events.rs:
```diff
@@ -858,6 +882,15 @@ impl Writeable for Event {
+				write_tlv_fields!(writer, {
+					(0, channel_id, required),
+					(2, user_channel_id, required),
+					(4, Some(counterparty_node_id), required),
```
Done in https://github.com/lightningdevkit/rust-lightning/pull/1743/commits/2f10adf05e148e045e78a99e706cf9438788e46d.

tnull commented:
> I think this is good - can you squash the fixups into the appropriate commits? Its somewhat hard to re-review when the fixups are all at the end rather than with the commit they'll eventually get squashed into.

Think all fixups belonged to the first commit => Done.

tnull commented:
Squashed fixups and included the pending changelog message

```
> git diff-tree -U2 a8720b2 49dfcb6
diff --git a/pending_changelog/1743.txt b/pending_changelog/1743.txt
new file mode 100644
index 00000000..93e41202
--- /dev/null
+++ b/pending_changelog/1743.txt
@@ -0,0 +1,7 @@
+## API Updates
+- A new `ChannelReady` event is generated whenever a channel becomes ready to
+  be used, i.e., after both sides sent the `channel_ready` message.
+
+## Backwards Compatibilty
+- No `ChannelReady` events will be generated for previously existing channels, including
+  those which become ready after upgrading 0.0.113.
```
